crates/matrix-sdk/src/send_queue/upload.rs

```diff
@@ -308,11 +254,10 @@ impl QueueStorage {
                 .map_err(RoomSendQueueStorageError::EventCacheStoreError)?;
 
             // Rename the thumbnail too, if needs be.
-            if let Some((info, new_source)) =
-                thumbnail_info.as_ref().zip(sent_media.thumbnail.clone())
+            if let Some((txn, new_source)) =
+                thumbnail_upload_txn.as_ref().zip(sent_media.thumbnail.clone())
             {
-                let from_req =
-                    make_local_thumbnail_media_request(&info.txn, info.height, info.width);
+                let from_req = Media::make_local_file_media_request(txn);
```

Comment:
Are we sure we want to do this? This cached the thumbnail with the exact same dimensions the user uploaded, which is actually nice. With this change, the media stored in the cache is renamed with an incorrect format (File, for a thumbnail), and it's unclear for external users that they can request the thumbnail as a file.

Reply:
1. It is not an "incorrect" format. What is important to remember here is that the format reflects only the request that would be made to the homeserver. There is no link between the format of the files stored in the media cache and their purpose in an event.

    If we were to use the `Thumbnail` format, it should reflect that it is a thumbnail of the original file of the event. The way that it is stored actually means that it is a thumbnail of the thumbnail file of the event.
    
    To clarify, with the Rust types:

    ```rust
    // What we have currently.
    // This is requesting a server-generated thumbnail of the thumbnail file
    // at the same dimensions as the thumbnail file.
    let request_thumbnail_of_thumbnail_file = MediaRequestParameters {
      source: thumbnail_file_source,
      format: MediaFormat::Thumbnail(settings_from_thumbnail_metadata),
    };

    // If we wanted to use a `Thumbnail` format, this would make more sense, imo.
    // This is requesting a server-generated thumbnail of the original file
    // at the same dimensions as the thumbnail file.
    let request_thumbnail_of_original_file = MediaRequestParameters {
      source: original_file_source,
      format: MediaFormat::Thumbnail(settings_from_thumbnail_metadata),
    };
    
    // This is correct, we are requesting the thumbnail file to the server.
    let request_thumbnail_file = MediaRequestParameters {
      source: thumbnail_file_source,
      format: MediaFormat::File,
    };
    ```
2. Clients might not request the exact dimensions of the thumbnail. Given that the API allows to request a thumbnail the fits inside given dimensions with `scale`, the client might not bother at all with the original dimensions of the thumbnail and always request, for example, a thumbnail that fits inside a `600x400` rectangle.

    So the need to request _exactly_ the dimensions of the thumbnail to hit the cache for locally-sent thumbnails is already something that needs to be documented for clients.
3. Clients **cannot** request a server-generated thumbnail of a file that is encrypted, it would be a wasteful request that always returns a 404. Therefore a request with the `Thumbnail` format should **never** be made in an encrypted room. Clients can only request the original file or the thumbnail file, so the format is always `File`, which means that the file that was cached will never be accessed.

    With the current code that uses the `Thumbnail` format, if a client wants to hit the media cache for locally-sent events in an encrypted room, they need special code to know if the media was sent from the client, and then only in this case try to request a server-generated thumbnail of the thumbnail file with the thumbnail file's dimensions. Or they don't treat encrypted and non-encrypted media differently and always make a wasteful request.
4. Since we enforce that thumbnails that were sent by the SDK provide all their metadata, a client should be able to request the thumbnail file instead of a server-generated thumbnail in all rooms if it matches the dimensions that they want.

Given all that, storing the thumbnail with a `File` format is less problematic imo.

Since points 3 and 4 mention logic that needs to be implemented in every client, maybe we can do it in the SDK, by improving the code of [`Media::get_thumbnail()`](https://matrix-org.github.io/matrix-rust-sdk/matrix_sdk/media/struct.Media.html#method.get_thumbnail) (which is subpar currently anyway).

Reply:
Ah yeah, that makes sense, I hadn't realized about the first point :facepalm: Bonus points if we document all of this super clearly/cleanly as part of the `Media::get_thumbnail()` refactor.